subsys/net/lib/conn_mgr/conn_mgr.c

```diff
@@ -95,52 +103,49 @@ static void conn_mgr_act_on_changes(void)
 	k_mutex_lock(&conn_mgr_lock, K_FOREVER);
 
 	for (idx = 0; idx < ARRAY_SIZE(iface_states); idx++) {
-		enum conn_mgr_state state;
+		bool is_ip_ready;
+		bool is_l4_ready;
+		bool was_l4_ready;
 
 		if (iface_states[idx] == 0) {
 			/* This interface is not used */
 			continue;
 		}
 
 		if (!(iface_states[idx] & CMGR_IF_EVT_CHANGED)) {
+			/* No changes on this iface */
 			continue;
 		}
 
-		state = CMGR_STATE_CONNECTED;
-
-		state &= conn_mgr_iface_status(idx);
-		if (state) {
-			enum conn_mgr_state ip_state =
-				CMGR_STATE_DISCONNECTED;
-
-			if (IS_ENABLED(CONFIG_NET_IPV6)) {
-				ip_state |= conn_mgr_ipv6_status(idx);
-			}
-
-			if (IS_ENABLED(CONFIG_NET_IPV4)) {
-				ip_state |= conn_mgr_ipv4_status(idx);
-			}
-
-			state &= ip_state;
-		}
-
+		/* Clear the state-change flag */
 		iface_states[idx] &= ~CMGR_IF_EVT_CHANGED;
 
-		if (state == CMGR_STATE_CONNECTED &&
-		    !(iface_states[idx] & CMGR_IF_STATE_READY)) {
-			iface_states[idx] |= CMGR_IF_STATE_READY;
+		/* Detect whether the iface is currently or was L4 ready */
+		is_ip_ready =	conn_mgr_ipv6_status(idx) || conn_mgr_ipv4_status(idx);
+		is_l4_ready =	conn_mgr_iface_is_oper_up(idx) && is_ip_ready;
+		was_l4_ready =	iface_states[idx] & CMGR_IF_STATE_READY;
 
-			conn_mgr_notify_status(idx);
-		} else if (state != CMGR_STATE_CONNECTED &&
-			   (iface_states[idx] & CMGR_IF_STATE_READY)) {
+		/* Respond to changes to iface readiness */
+		if (was_l4_ready != is_l4_ready) {
+			/* Clear and then update the L4-readiness bit */
 			iface_states[idx] &= ~CMGR_IF_STATE_READY;
 
-			conn_mgr_notify_status(idx);
+			if (is_l4_ready) {
+				iface_states[idx] |= CMGR_IF_STATE_READY;
+			}
+
+			/* Notify listeners of the readiness change */
+			conn_mgr_notify_if_readiness(idx);
```

Comment:
Yep, that's right, the goal here was to clean up the function without changing its functionality. I had two primary reasons for doing this:
- It makes it easier to write tests, since the operation of conn_mgr is (to me, at least) a bit easier to follow
- It makes adding the global connectivity counting feature you mention significantly less ugly to implement